File: tests/python/test_capabilities_smoke.py
"""Capability smoke tests: streaming, structured output, agent tool calling.

Gated on the OPENROUTER_API_KEY environment variable.
"""

import json
import os

import pytest

from blazen import AgentResult, ChatMessage, CompletionModel, CompletionOptions, ToolDef, run_agent

OPENROUTER_API_KEY = os.environ.get("OPENROUTER_API_KEY")

skip_without_key = pytest.mark.skipif(
    not OPENROUTER_API_KEY,
    reason="OPENROUTER_API_KEY not set",
)


# ---------------------------------------------------------------------------
# Streaming
# ---------------------------------------------------------------------------


@skip_without_key
@pytest.mark.asyncio
async def test_streaming_completion():
    """Stream a completion and verify at least one chunk is received."""
    model = CompletionModel.openrouter(OPENROUTER_API_KEY)

    chunks: list[dict] = []

    def on_chunk(chunk: dict) -> None:
        chunks.append(chunk)

    await model.stream(
        [ChatMessage.user("Count from 1 to 5.")],
        on_chunk,
        CompletionOptions(max_tokens=64),
    )

    assert len(chunks) > 0, "Expected at least one streamed chunk"


# ---------------------------------------------------------------------------
# Structured output (JSON schema)
# ---------------------------------------------------------------------------


@skip_without_key
@pytest.mark.asyncio
async def test_structured_output():
    """Request structured JSON output and verify the schema is respected."""
    model = CompletionModel.openrouter(OPENROUTER_API_KEY)

    response = await model.complete(
        [ChatMessage.user("What is 2+2? Return JSON.")],
        CompletionOptions(
            response_format={
                "type": "json_schema",
                "json_schema": {
                    "name": "math",
                    "strict": True,
                    "schema": {
                        "type": "object",
                        "properties": {
                            "answer": {"type": "integer"},
                        },
                        "required": ["answer"],
                    },
                },
            },
            max_tokens=64,
        ),
    )

    assert response.content is not None
    parsed = json.loads(response.content)
    assert parsed["answer"] == 4


# ---------------------------------------------------------------------------
# Agent tool calling (sync handler)
# ---------------------------------------------------------------------------


@skip_without_key
@pytest.mark.asyncio
async def test_agent_tool_calling():
    """Run the agent loop with a sync tool handler."""
    model = CompletionModel.openrouter(OPENROUTER_API_KEY)

    def multiply(arguments: dict) -> dict:
        return {"result": arguments["a"] * arguments["b"]}

    tool = ToolDef(
        name="multiply",
        description="Multiply two numbers together.",
        parameters={
            "type": "object",
            "properties": {
                "a": {"type": "number"},
                "b": {"type": "number"},
            },
            "required": ["a", "b"],
        },
        handler=multiply,
    )

    result = await run_agent(
        model,
        [ChatMessage.user("What is 15 * 7? Use the multiply tool.")],
        tools=[tool],
        max_iterations=5,
    )

    assert isinstance(result, AgentResult)
    assert result.response.content is not None
    assert "105" in result.response.content
    assert result.iterations >= 1


# ---------------------------------------------------------------------------
# Agent tool calling (async handler)
# ---------------------------------------------------------------------------


@skip_without_key
@pytest.mark.asyncio
async def test_agent_async_tool():
    """Run the agent loop with an async tool handler."""
    model = CompletionModel.openrouter(OPENROUTER_API_KEY)

    async def multiply(arguments: dict) -> dict:
        return {"result": arguments["a"] * arguments["b"]}

    tool = ToolDef(
        name="multiply",
        description="Multiply two numbers together.",
        parameters={
            "type": "object",
            "properties": {
                "a": {"type": "number"},
                "b": {"type": "number"},
            },
            "required": ["a", "b"],
        },
        handler=multiply,
    )

    result = await run_agent(
        model,
        [ChatMessage.user("What is 15 * 7? Use the multiply tool.")],
        tools=[tool],
        max_iterations=5,
    )

    assert isinstance(result, AgentResult)
    assert result.response.content is not None
    assert "105" in result.response.content
    assert result.iterations >= 1

File: tests/node/test_capabilities_smoke.mjs
/**
 * Capabilities smoke tests: streaming, structured output, and agent tool calling.
 *
 * Gated on the OPENROUTER_API_KEY environment variable.
 * Only runs during release CI or manual invocation.
 */

import test from "ava";

import {
  CompletionModel,
  ChatMessage,
  runAgent,
} from "../../crates/blazen-node/index.js";

const OPENROUTER_API_KEY = process.env.OPENROUTER_API_KEY;
const T = OPENROUTER_API_KEY ? test : test.skip;

// -- Streaming ---------------------------------------------------------------

T("Streaming smoke tests · streams chunks for a basic prompt", async (t) => {
  const model = CompletionModel.openrouter({ apiKey: OPENROUTER_API_KEY });
  const chunks = [];

  await model.stream(
    [ChatMessage.user("Count from 1 to 5.")],
    (chunk) => {
      chunks.push(chunk);
    }
  );

  t.truthy(chunks.length > 0, "expected at least one streamed chunk");

  // At least some chunks should carry a delta with text content
  const hasTextDelta = chunks.some(
    (c) => typeof c.delta === "string" && c.delta.length > 0
  );
  t.truthy(hasTextDelta, "expected at least one chunk with a non-empty delta");
});

// -- Structured Output -------------------------------------------------------

T("Structured output smoke tests · returns JSON conforming to a schema", async (t) => {
  const model = CompletionModel.openrouter({ apiKey: OPENROUTER_API_KEY });

  const response = await model.completeWithOptions(
    [ChatMessage.user("What is 2 + 2? Respond in the required JSON format.")],
    {
      responseFormat: {
        type: "json_schema",
        json_schema: {
          name: "math",
          strict: true,
          schema: {
            type: "object",
            properties: {
              answer: { type: "integer" },
            },
            required: ["answer"],
          },
        },
      },
    }
  );

  t.truthy(response.content, "expected content in response");

  const parsed = JSON.parse(response.content);
  t.is(parsed.answer, 4, `expected answer=4, got ${parsed.answer}`);
});

// -- Agent Tool Calling ------------------------------------------------------

T("Agent tool calling smoke tests · uses a multiply tool and returns the correct result", async (t) => {
  const model = CompletionModel.openrouter({ apiKey: OPENROUTER_API_KEY });

  const tools = [
    {
      name: "multiply",
      description: "Multiply two numbers together and return the product.",
      parameters: {
        type: "object",
        properties: {
          a: { type: "number", description: "First number" },
          b: { type: "number", description: "Second number" },
        },
        required: ["a", "b"],
      },
    },
  ];

  const toolHandler = async (toolName, args) => {
    if (toolName === "multiply") {
      return JSON.stringify({ result: args.a * args.b });
    }
    throw new Error(`Unknown tool: ${toolName}`);
  };

  const result = await runAgent(
    model,
    [ChatMessage.user("What is 15 * 7? Use the multiply tool.")],
    tools,
    toolHandler,
    { maxIterations: 5 }
  );

  t.truthy(result.response, "expected response in agent result");
  t.truthy(result.response.content, "expected content in agent response");
  t.truthy(
    result.response.content.includes("105"),
    `expected '105' in: ${result.response.content}`
  );
  t.truthy(
    result.iterations >= 1,
    `expected at least 1 iteration, got ${result.iterations}`
  );
});
